### PR TITLE
 Fix: fix issue 842 for udt holders count

### PR DESCRIPTION
**Pull request by Naupio**

### CODE REVIEW

File: lib/godwoken_explorer/udt.ex [hunk 1]
```diff
@@ -75,9 +75,20 @@ defmodule GodwokenExplorer.UDT do
     end
   end
 
-  def count_holder(udt_id) do
-    from(cbub in CurrentBridgedUDTBalance, where: cbub.udt_id == ^udt_id)
-    |> Repo.aggregate(:count)
+  def count_holder(udt) do
```
Comment (Keith-CY) [inline lib/godwoken_explorer/udt.ex:78]: If there are a bridged token of id `1` and its mapped native token of id `2`, will they return different values?

File: test/graphql/udt_test.exs [hunk 2]
```diff
@@ -139,4 +139,48 @@ defmodule GodwokenExplorer.Graphql.UDTTest do
              json_response(conn, 200)
            )
   end
+
+  test "graphql: udt holders ", %{
```
Comment (Keith-CY) [inline test/graphql/udt_test.exs:143]: update on native token is triggered by its bridged one should be covered in the test

File: lib/godwoken_explorer/udt.ex (hunk 1 above)
Comment (zmcNotafraid) [reply lib/godwoken_explorer/udt.ex:78]: I think is better to return same value for bridge udt.

File: test/graphql/udt_test.exs (hunk 2 above)
Comment (Naupio) [reply test/graphql/udt_test.exs]: Has been added

File: lib/godwoken_explorer/udt.ex (hunk 1 above)
Comment (Naupio) [reply lib/godwoken_explorer/udt.ex:78]: so, does  bridged token should return it's mapping native token `holders count` ? @Keith-CY @zmcNotafraid

**Review by Keith-CY — APPROVED** [review]

**Comment by Keith-CY** [comment]
We'll sync other data of native token and bridge token point to a same asset in another PR

File: lib/godwoken_explorer/udt.ex [hunk 3]
```diff
@@ -75,9 +77,68 @@ defmodule GodwokenExplorer.UDT do
...
+    case udt.type do
+      :native ->
+        bridge_udt_query = from(u in UDT, where: u.bridge_account_id == ^udt.id)
+
+        case Repo.one(bridge_udt_query) do
+          nil ->
+            from(cub in CurrentUDTBalance,
+              where: cub.token_contract_address_hash == ^udt.contract_address_hash
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/udt.ex:88]: here need to filter value is 0 records

File: lib/godwoken_explorer/udt.ex [hunk 4]
```diff
@@ -75,9 +77,68 @@ defmodule GodwokenExplorer.UDT do
...
+            )
+            |> Repo.aggregate(:count)
+
+          %UDT{id: id} ->
+            cu_query =
+              from(cub in CurrentUDTBalance,
+                where: cub.token_contract_address_hash == ^udt.contract_address_hash
+              )
+              |> join(:inner, [cub], u in UDT,
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/udt.ex:97]: here not need for join udt, because you already know a udt

File: lib/godwoken_explorer/udt.ex [hunk 5]
```diff
@@ -75,9 +77,68 @@ defmodule GodwokenExplorer.UDT do
...
+                on: cub.token_contract_address_hash == u.contract_address_hash
+              )
+              |> select_merge([_, u], %{uniq_id: u.id})
+
+            cbu_query =
+              from(cbub in CurrentBridgedUDTBalance, where: cbub.udt_id == ^id)
+              |> join(:inner, [cbub], u in UDT, on: cbub.udt_id == u.id)
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/udt.ex:104]: why not uniq address_hash ?

**Review by zmcNotafraid — CHANGES_REQUESTED** [review]

File: lib/godwoken_explorer/udt.ex (hunk 3 above)
Comment (Naupio) [reply lib/godwoken_explorer/udt.ex]: fix here: https://github.com/Magickbase/godwoken_explorer/pull/849/files#diff-d44d00acb06eaf71b14ba5b0fb3233b779927732d67685af828afda5e67694af

File: lib/godwoken_explorer/udt.ex (hunk 5 above)
Comment (Naupio) [reply lib/godwoken_explorer/udt.ex]: fix here
https://github.com/Magickbase/godwoken_explorer/pull/849/files#diff-d44d00acb06eaf71b14ba5b0fb3233b779927732d67685af828afda5e67694af

File: lib/godwoken_explorer/udt.ex (hunk 4 above)
Comment (Naupio) [reply lib/godwoken_explorer/udt.ex]: fix here
https://github.com/Magickbase/godwoken_explorer/pull/849/files#diff-d44d00acb06eaf71b14ba5b0fb3233b779927732d67685af828afda5e67694af

**Comment by Naupio** [comment]
improve with this commit: https://github.com/Magickbase/godwoken_explorer/pull/849/commits/e0ed14ac2f0e23fd6a24322948925562d652014c

pls review the latest code @zmcNotafraid

File: lib/godwoken_explorer/udt.ex [hunk 6]
```diff
@@ -75,8 +77,76 @@ defmodule GodwokenExplorer.UDT do
...
+          [c],
+          c.udt_id == ^udt.id and c.value > 0
+        )
+      else
+        false
+      end
+    end
+  end
+
+  defmacrop dyn_mapping_udt_condtion(udt) do
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/udt.ex:106]: 这个方法怎么理解？

File: lib/godwoken_explorer/udt.ex [hunk 7]
```diff
@@ -75,8 +77,76 @@ defmodule GodwokenExplorer.UDT do
...
+    end
+  end
+
+  def count_holder(udt) do
+    mapping_udt_conditions = dyn_mapping_udt_condtion(udt)
+    mapping_udt_query = from(u in UDT, where: ^mapping_udt_conditions)
+    mapping_udt = Repo.one(mapping_udt_query)
+
+    cub1_conditions = dyn_cub_condition(udt)
+    cub2_conditions = dyn_cub_condition(mapping_udt)
```
Comment (zmcNotafraid) [inline lib/godwoken_explorer/udt.ex:129]: 这个方法的参数有可能是nil 把？

Comment (Naupio) [reply lib/godwoken_explorer/udt.ex:145]: fixed

**Review by zmcNotafraid — APPROVED** [review]

**Pull request merged**